### PR TITLE
Garbage-collect network stack

### PR DESCRIPTION
**Pull request by pcercuei**
Tweak the code so that the network stack (and BBA and LAN adapter code) will be garbage-collected when a KOS program has been compiled without the INIT_NET flag.

This really drops the size of the compiled binaries. For instance, the dreamcast "hello" example dropped from 284 KiB to 156 KiB (both stripped).

Note that this breaks the ABI a bit, as KOS' symbols are not exported anymore by default, and will be exported only if the `INIT_EXPORT` flag has been passed to `KOS_INIT_FLAGS`. This should be fine for the vast majority of programs out there though.

### CODE REVIEW

File: kernel/arch/dreamcast/hardware/hardware.c [hunk 1]
```diff
@@ -40,6 +40,22 @@ int hardware_sys_init(void) {
     return 0;
 }
 
+void (*bba_la_init_weak)(void) __attribute__((weak));
+void (*bba_la_shutdown_weak)(void) __attribute__((weak));
+
+void bba_la_init(void)
+{
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/hardware.c:47]:
```suggestion
void bba_la_init(void) {
```

File: kernel/arch/dreamcast/hardware/hardware.c [hunk 2]
```diff
@@ -40,6 +40,22 @@ int hardware_sys_init(void) {
...
+    /* Setup network (this won't do anything unless we enable netcore) */
+    bba_init();
+    la_init();
+}
+
+void bba_la_shutdown(void)
+{
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/hardware.c:54]:
```suggestion
void bba_la_shutdown(void) {
```

File: kernel/arch/dreamcast/hardware/hardware.c [hunk 3]
```diff
@@ -57,9 +73,8 @@ int hardware_periph_init(void) {
     vid_init(DEFAULT_VID_MODE, DEFAULT_PIXEL_MODE);
 
 #ifndef _arch_sub_naomi
-    /* Setup network (this won't do anything unless we enable netcore) */
-    bba_init();
-    la_init();
+    if (bba_la_init_weak)
+	    (*bba_la_init_weak)();
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/hardware.c:77]:
```suggestion
    if(bba_la_init_weak)
        (*bba_la_init_weak)();
```

File: kernel/arch/dreamcast/hardware/hardware.c [hunk 4]
```diff
@@ -71,8 +86,8 @@ void hardware_shutdown(void) {
     switch(initted) {
         case 2:
 #ifndef _arch_sub_naomi
-            la_shutdown();
-            bba_shutdown();
+            if (bba_la_shutdown_weak)
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/hardware.c:89]:
```suggestion
            if(bba_la_shutdown_weak)
```

File: kernel/arch/dreamcast/kernel/init.c [hunk 5]
```diff
@@ -59,17 +59,41 @@ dbgio_handler_t * dbgio_handlers[] = {
 };
 int dbgio_handler_cnt = sizeof(dbgio_handlers) / sizeof(dbgio_handler_t *);
 
-/* Auto-init stuff: override with a non-weak symbol if you don't want all of
-   this to be linked into your code (and do the same with the
-   arch_auto_shutdown function too). */
-int  __attribute__((weak)) arch_auto_init(void) {
-#ifndef _arch_sub_naomi
+void arch_init_net(void)
+{
```
Comment (ljsebald) [inline kernel/arch/dreamcast/kernel/init.c:63]:
```suggestion
void arch_init_net(void) {
```

File: kernel/arch/dreamcast/kernel/init.c [hunk 6]
```diff
@@ -155,31 +179,8 @@ int  __attribute__((weak)) arch_auto_init(void) {
...
-            fs_dclsocket_init_console();
-
-            if(!fs_dclsocket_init()) {
-                dbgio_dev_select("fs_dclsocket");
-                dbgio_enable();
-                dbglog(DBG_INFO, "fs_dclsocket console support enabled\n");
-            }
-        }
-    }
+    if (init_net_weak)
```
Comment (ljsebald) [inline kernel/arch/dreamcast/kernel/init.c:182]:
```suggestion
    if(init_net_weak)
```

File: kernel/arch/dreamcast/kernel/init.c [hunk 7]
```diff
@@ -188,7 +189,8 @@ int  __attribute__((weak)) arch_auto_init(void) {
 void  __attribute__((weak)) arch_auto_shutdown(void) {
 #ifndef _arch_sub_naomi
     fs_dclsocket_shutdown();
-    net_shutdown();
+    if (net_shutdown_weak)
+	    (*net_shutdown_weak)();
```
Comment (ljsebald) [inline kernel/arch/dreamcast/kernel/init.c:193]:
```suggestion
    if(net_shutdown_weak)
        (*net_shutdown_weak)();
```

File: kernel/exports/nmmgr.c [hunk 8]
```diff
@@ -96,7 +98,8 @@ int nmmgr_init(void) {
     LIST_INIT(&nmmgr_handlers);
 
     /* Initialize our internal exports */
-    export_init();
+    if (export_init_weak)
+	    (*export_init_weak)();
```
Comment (ljsebald) [inline kernel/exports/nmmgr.c:102]:
```suggestion
    if(export_init_weak)
        (*export_init_weak)();
```

File: include/kos/init.h [hunk 9]
```diff
@@ -38,7 +38,18 @@ __BEGIN_DECLS
...
+	extern void arch_init_net(void); \
+	void (*init_net_weak)(void) = ((flags) & INIT_NET) ? arch_init_net : NULL; \
+	extern void net_shutdown(void); \
+	void (*net_shutdown_weak)(void) = ((flags) & INIT_NET) ? net_shutdown : NULL; \
+	extern void bba_la_init(void); \
+	void (*bba_la_init_weak)(void) = ((flags) & INIT_NET) ? bba_la_init : NULL; \
+	extern void bba_la_shutdown(void); \
+	void (*bba_la_shutdown_weak)(void) = ((flags) & INIT_NET) ? bba_la_shutdown : NULL; \
+	extern int export_init(void); \
+	int (*export_init_weak)(void) = ((flags) & INIT_EXPORT) ? export_init : NULL
```
Comment (ljsebald) [inline include/kos/init.h:52]:
```suggestion
    uint32 __kos_init_flags = (flags); \
    extern void arch_init_net(void); \
    void (*init_net_weak)(void) = ((flags) & INIT_NET) ? arch_init_net : NULL; \
    extern void net_shutdown(void); \
    void (*net_shutdown_weak)(void) = ((flags) & INIT_NET) ? net_shutdown : NULL; \
    extern void bba_la_init(void); \
    void (*bba_la_init_weak)(void) = ((flags) & INIT_NET) ? bba_la_init : NULL; \
    extern void bba_la_shutdown(void); \
    void (*bba_la_shutdown_weak)(void) = ((flags) & INIT_NET) ? bba_la_shutdown : NULL; \
    extern int export_init(void); \
    int (*export_init_weak)(void) = ((flags) & INIT_EXPORT) ? export_init : NULL
```

**Review by ljsebald — CHANGES_REQUESTED** [review]
Just have a few minor style fixes here.

Thank you for taking this on. This is something that's actually been on my mind for a while, but that I hadn't sat down and worked on. Definitely an improvement to cut down on the filesize in this way.

**Comment by pcercuei** [comment]
@ljsebald updated.

File: kernel/exports/nmmgr.c [hunk 10]
```diff
@@ -96,7 +98,8 @@ int nmmgr_init(void) {
     LIST_INIT(&nmmgr_handlers);
 
     /* Initialize our internal exports */
-    export_init();
+    if (export_init_weak)
```
Comment (ljsebald) [inline kernel/exports/nmmgr.c:101]:
```suggestion
    if(export_init_weak)
```

**Review by ljsebald — DISMISSED** [review]

**Pull request merged**